### PR TITLE
Busco

### PR DESCRIPTION
**Pull request by priyanka-surana**
Gets possible odb lineages from GOAT, picks the best choice, runs BUSCO, and creates a csv table.

> **Please check the logic in `bin/get_odb.py`**

## PR checklist

- [ x] This comment contains a description of changes (with reason).
- [x ] If you've fixed a bug or added code that should be tested, add tests!
  - [ x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/genomenote/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/genomenote _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x ] Make sure your code lints (`nf-core lint`).
- [x ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

**Review by muffato — COMMENTED** [review]
🤩 I love it ! It'm truly in awe, seeing the test profile run and complete within minutes. Well done !

File: conf/test.config [hunk 1]
```diff
@@ -21,10 +21,7 @@ params {
...
     input   = "mMelMel3"
-    project = ".sandbox_ps22"
-    kmer = "/lustre/scratch124/tol/projects/darwin/data/fungi/Laetiporus_sulphureus/genomic_data/gfLaeSulp1/pacbio/kmer/k31/"
-    //input = "/lustre/scratch124/tol/projects/.sandbox_ps22/data/mammals/Meles_meles/analysis/mMelMel3.2_paternal_haplotype/testing/samplesheet.csv"
-
-    // Genome references
-    //fasta = "/lustre/scratch124/tol/projects/.sandbox_ps22/data/mammals/Meles_meles/analysis/mMelMel3.2_paternal_haplotype/assembly/indices/GCA_922984935.2.subset.unmasked.fasta"
+    project = ".sandbox"
+    kmer = "/lustre/scratch123/tol/projects/.sandbox/data/mammals/Meles_meles/genomic_data/mMelMel1/illumina/kmer/k31"
+    lineage_db = "/lustre/scratch123/tol/teams/blaxter/users/sk13/busco_2021_06_reduced"
```
[the remark this answers is not in the document]
Comment (priyanka-surana) [reply conf/test.config:26]: Done 👍

File: modules/local/samtools_view.nf [hunk 2]
```diff
@@ -1,36 +1,44 @@
...
 
     output:
-    tuple val(meta), path("*.bam") , emit: bam
-    path("*.fai")                  , emit: fai
-    path  "versions.yml"           , emit: versions
+    tuple val(meta), path("*.bam"),  emit: bam,  optional: true
+    tuple val(meta), path("*.cram"), emit: cram, optional: true
+    tuple val(meta), path("*.sam"),  emit: sam,  optional: true
+    tuple val(meta), path("*.csi"),  emit: csi,  optional: true
+    path("*.fai"),                   emit: fai
```
[the remark this answers is not in the document]
Comment (priyanka-surana) [reply modules/local/samtools_view.nf:19]: No we leave `samtools_view` as it is and will use `samtools_faidx` for `fai`. Will update accordingly.

File: subworkflows/local/genome_statistics.nf [hunk 3]
```diff
@@ -3,22 +3,38 @@
...
+include { GET_ODB             } from '../../modules/local/get_odb'
+include { BUSCO               } from '../../modules/nf-core/modules/busco/main'
 include { CREATE_TABLE        } from '../../modules/local/create_table'
 
 workflow GENOME_STATISTICS {
     take:
-    asm
+    genome
+    lineage_db
     kmer
```
[the remark this answers is not in the document]
Comment (priyanka-surana) [reply subworkflows/local/genome_statistics.nf:14]: Will do 👍

File: bin/get_odb.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    return parser.parse_args(args)
+
+def make_dir(path):
+    if len(path) > 0:
+        os.makedirs(path, exist_ok=True)
+
+def get_odb(asm, file_out):
+    response = requests.get("https://goat.genomehubs.org/api/v2/search?query=tax_lineage%28queryA.taxon_id%29&queryA=assembly--assembly_id%3D"+asm+"&result=taxon&fields=odb10_lineage").json()["results"]
+    odb_arr = [response[x]["result"]["fields"]["odb10_lineage"]["value"] for x in range(len(re.findall("result", json.dumps(response))))]
+    odb_val = "eutheria_odb10" if "eutheria_odb10" in odb_arr else odb_arr[0]
```
[the remark this answers is not in the document]
Comment (priyanka-surana) [reply bin/get_odb.py:26]: I can email Steph or if we have the new genome editor I can talk to them.

[the remark this answers is not in the document]
Comment (priyanka-surana) [reply bin/get_odb.py:26]: And will add further documentation

File: bin/tol_input.sh [hunk 5]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-PROJECT_BASEDIR=/lustre/scratch124/tol/projects
+PROJECT_BASEDIR=/lustre/scratch123/tol/projects
```
[the remark this answers is not in the document]
Comment (priyanka-surana) [reply bin/tol_input.sh:3]: Implemented in this pipeline. Take a look.

File: modules/local/samtools_view.nf (hunk 2 above)
[the remark this answers is not in the document]
Comment (priyanka-surana) [reply modules/local/samtools_view.nf:19]: Done

File: bin/get_odb.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    parser = argparse.ArgumentParser(description=Description)
+    parser.add_argument("ASSEMBLY", help="Assembly GCA value.")
+    parser.add_argument("FILE_OUT", help="Output CSV file.")
+    parser.add_argument('--version', action='version', version='%(prog)s 1.0')
+    return parser.parse_args(args)
+
+def make_dir(path):
+    if len(path) > 0:
+        os.makedirs(path, exist_ok=True)
+
```
[the remark this answers is not in the document]
Comment (priyanka-surana) [reply bin/get_odb.py:22]: Used now

File: bin/get_odb.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+def get_odb(asm, file_out):
+    response = requests.get("https://goat.genomehubs.org/api/v2/search?query=tax_lineage%28queryA.taxon_id%29&queryA=assembly--assembly_id%3D"+asm+"&result=taxon&fields=odb10_lineage").json()["results"]
+    odb_arr = [response[x]["result"]["fields"]["odb10_lineage"]["value"] for x in range(len(re.findall("result", json.dumps(response))))]
```
[the remark this answers is not in the document]
Comment (priyanka-surana) [reply bin/get_odb.py:25]: Changed

File: modules/local/get_odb.nf [hunk 8]
```diff
@@ -0,0 +1,27 @@
+process GET_ODB {
+    tag "${meta.id}"
+    label 'process_single'
+
+    conda (params.enable_conda ? "conda-forge::requests=2.28.1" : null)
```
[the remark this answers is not in the document]
Comment (priyanka-surana) [reply modules/local/get_odb.nf:5]: Updated

File: bin/create_table.py [hunk 9]
```diff
@@ -30,9 +31,21 @@ def extract_n50(file_in, file_out):
...
+def extract_busco(file_in, file_out):
+    with open(file_in, "r") as fin:
+        data = json.load(fin)
+
+    out_dir = os.path.dirname(file_out)
+    make_dir(out_dir)
+
+    with open(file_out, "a") as fout:
+        print("busco_lineage", data["lineage_dataset"]["name"], sep=",", file=fout)
+        print("busco_summary", data["results"]["one_line_summary"], sep=",", file=fout)
```
[the remark this answers is not in the document]
Comment (priyanka-surana) [reply bin/create_table.py:43]: Updated

**Comment by priyanka-surana** [comment]
Testing works successfully. Same command as before in interactive session:
```
nextflow run main.nf -profile test,singularity -c ../genomenote/singularity_runOptions.config
```

File: conf/test_full.config [hunk 10]
```diff
@@ -15,10 +15,9 @@ params {
     config_profile_description = 'Full test dataset to check pipeline function'
 
     // Input data for full size test
-    input   = "gfLaeSulp1"
-    project = ".sandbox_ps22"
-    //input = "/lustre/scratch124/tol/projects/.sandbox_ps22/data/fungi/Laetiporus_sulphureus/analysis/gfLaeSulp1.1/testing/samplesheet.csv"
+    input      = "gfLaeSulp1"
+    project    = "/lustre/scratch123/tol/projects/.sandbox"
+    kmer       = "/lustre/scratch123/tol/projects/.sandbox/data/fungi/Laetiporus_sulphureus/genomic_data/gfLaeSulp1/hic-arima2/kmer"
+    lineage_db = ""
```
Comment (muffato) [inline conf/test_full.config:21]:
```suggestion
    lineage_db = "/lustre/scratch123/tol/resources/busco/v5"
```

**Review by muffato — APPROVED** [review]

File: conf/test_full.config (hunk 10 above)
Comment (priyanka-surana) [reply conf/test_full.config]: Didn't know about this. Thank you 😊

File: modules/local/get_odb.nf [hunk 11]
```diff
@@ -0,0 +1,27 @@
+process GET_ODB {
+    tag "${meta.id}"
+    label 'process_single'
+
+    conda (params.enable_conda ? "conda-forge::requests=2.26.0" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/requests:2.26.0' :
+        'quay.io/biocontainers/requests:2.26.0' }"
+
```
Comment (muffato) [inline modules/local/get_odb.nf:9]: This may appease the EditorConfig check ?
```suggestion

```

**Pull request merged**